### PR TITLE
Add ToGray preprocessing layers

### PR DESCRIPTION
**Pull request by innat**
make-up https://github.com/keras-team/keras-cv/issues/120

```
ToGray(num_output_channels=3)
```

```
ToGray(num_output_channels=1)
```

### CODE REVIEW

File: keras_cv/layers/preprocessing/togray.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                f"Valid arugment for num_output_channels param are 1 or 3. Got {num_output_channels}"
+            )
+        self.num_output_channels = num_output_channels
+
+    def _rgb_to_grayscale(self, image):
+        if self.num_output_channels == 1:
+            return tf.image.rgb_to_grayscale(image)
+        elif self.num_output_channels == 3:
+            _grayscale = tf.image.rgb_to_grayscale(image)
+            return tf.concat([_grayscale, _grayscale, _grayscale], axis=-1)
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/togray.py:56]: What's the use case for 3 channel grayscale?

**Review by LukeWood — CHANGES_REQUESTED** [review]

File: keras_cv/layers/preprocessing/togray.py (hunk 1 above)
Comment (innat) [reply keras_cv/layers/preprocessing/togray.py:56]: Let's say you want to train a model with RGB input and want to use this layer as an augmentation. Then you need to pass num_output_channels to 3. And similarly, for the case of grayscale input training, you will need to pass 1. FYI, `torch.transform` supports this way also.

File: keras_cv/layers/preprocessing/togray.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+import tensorflow.keras.layers as layers
+from tensorflow.keras import layers, backend
+
+
+class ToGray(layers.Layer):
```
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/layers/preprocessing/togray.py:20]: I chose this naming style inspired by `torch` and `albumentaiton`.  

If it's hard to decide the name, let's do that later.

File: keras_cv/layers/preprocessing/togray.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    def _check_input_params(self, num_output_channels):
+        if num_output_channels not in [1, 3]:
+            raise ValueError(
+                f"Valid arugment for num_output_channels param are 1 or 3. Got {num_output_channels}"
+            )
+        self.num_output_channels = num_output_channels
+
+    def _rgb_to_grayscale(self, image):
+        if self.num_output_channels == 1:
+            return tf.image.rgb_to_grayscale(image)
```
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/layers/preprocessing/togray.py:53]: Oh, I see. I think KPL will follow [0, 255] range. 

How about we just check if the input is within int or float and set feedback based on that. 

The `tf.image*` has lots of things KPL can get leveraged from.

File: keras_cv/layers/preprocessing/togray.py (hunk 1 above)
Comment (LukeWood) [reply keras_cv/layers/preprocessing/togray.py:60]: Good point, thanks.

File: keras_cv/layers/preprocessing/togray.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply keras_cv/layers/preprocessing/togray.py:20]: `GrayscaleConversion`? Or just `Grayscale`?

File: keras_cv/layers/preprocessing/togray.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bhack) [reply keras_cv/layers/preprocessing/togray.py:57]: I don't think that `tf.image` has any asumption about this in many ops e.g. in the grayscale conversion:

https://github.com/tensorflow/tensorflow/blob/v2.8.0/tensorflow/python/ops/image_ops_impl.py#L2484-L2517

[the remark this answers is not in the document]
Comment (bhack) [reply keras_cv/layers/preprocessing/togray.py:57]:
```
import tensorflow as tf
original = tf.constant([[[220.0, 125.0, 3.0]]])
print(original.dtype)
converted = tf.image.rgb_to_grayscale(original)
print(converted.numpy())
```

```
<dtype: 'float32'>
<dtype: 'float32'> [[[139.475]]]
```

File: keras_cv/layers/preprocessing/togray.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/layers/preprocessing/togray.py:20]: Yeah, I think `Grayscale` is better and more precise.

[the remark this answers is not in the document]
Comment (LukeWood) [reply keras_cv/layers/preprocessing/togray.py:20]: Cool, yeah lets use `GrayScale`.

File: keras_cv/layers/preprocessing/togray.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/layers/preprocessing/togray.py:57]: > I don't think that tf.image has any assumption about this in many ops e.g. in the grayscale conversion:

So, I guess in that case it now shouldn't be an issue.

File: keras_cv/layers/preprocessing/grayscale.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    Usage:
+    ```python
+    (images, labels), _ = tf.keras.datasets.cifar10.load_data()
+    to_grayscale = keras_cv.layers.preprocessing.ToGray()
+    augmented_images = to_grayscale(images)
+    ```
+    """
+
+    def __init__(self, num_output_channels=1, **kwargs):
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/grayscale.py:43]: lets rename `num_output_channels` to `output_channels`.

File: keras_cv/layers/preprocessing/grayscale.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    Args:
+        num_output_channels. It represent the output channel number after transformation. Valid values are 1 and 3.
+        (..., height, width, 1) for num_output_channels = 1 .
+        (..., height, width, 3) for num_output_channels = 3 .
+
+
+    Usage:
+    ```python
+    (images, labels), _ = tf.keras.datasets.cifar10.load_data()
+    to_grayscale = keras_cv.layers.preprocessing.ToGray()
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/grayscale.py:38]: Update docstring with the new name

File: keras_cv/layers/preprocessing/grayscale.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+import tensorflow.keras.layers as layers
+from tensorflow.keras import layers, backend
+
+
+class Grayscale(layers.Layer):
+    """Grayscale class for transforming RGB image to Grayscale image. The expected images
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/grayscale.py:21]: Update to:

Grayscale is a preprocessing layer that transforms RGB images to Grayscale images.  

Input images should have values in the range of [0, 255].

File: keras_cv/layers/preprocessing/grayscale.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    """Grayscale class for transforming RGB image to Grayscale image. The expected images 
+    should be [0-255] pixel ranges.
+    Input shape:
+        3D (unbatched) or 4D (batched) tensor with shape:
+        `(..., height, width, channels)`, in `"channels_last"` format
+    Output shape:
+        3D (unbatched) or 4D (batched) tensor with shape:
+        `(..., height, width, channels)`, in `"channels_last"` format
+    Args:
+        num_output_channels. It represent the output channel number after transformation. Valid values are 1 and 3.
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/grayscale.py:30]: remember to update this as well when you rename to `output_channels`

File: keras_cv/layers/preprocessing/grayscale.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        (..., height, width, 1) for num_output_channels = 1 .
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/grayscale.py:31]: explain these comments here.  I see what you're going for but be sure to clearly explain this.  Also indent one more time.

File: keras_cv/layers/preprocessing/grayscale.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+    def __init__(self, num_output_channels=1, **kwargs):
+        super().__init__(**kwargs)
+        self.num_output_channels = num_output_channels
+
+    def _check_input_params(self, num_output_channels):
+        if num_output_channels not in [1, 3]:
+            raise ValueError(
+                f"Valid arugment for num_output_channels param are 1 or 3. \
+                 1 for (..., height, width, 1) and 3 for (..., height, width, 3). Got {num_output_channels}"
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/grayscale.py:51]: Drop the sentence:
`1 for (..., height, width, 1) and 3 for (..., height, width, 3).`.  They can read the docstring.

File: keras_cv/layers/preprocessing/grayscale.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    """
+
+    def __init__(self, num_output_channels=1, **kwargs):
+        super().__init__(**kwargs)
+        self.num_output_channels = num_output_channels
+
+    def _check_input_params(self, num_output_channels):
+        if num_output_channels not in [1, 3]:
+            raise ValueError(
+                f"Valid arugment for num_output_channels param are 1 or 3. \
```
Comment (LukeWood) [inline keras_cv/layers/preprocessing/grayscale.py:50]: Received invalid argument `output_channels`.  `output_channels` must be in `1` or `3`.  Got ...

File: keras_cv/layers/preprocessing/grayscale.py (hunk 4 above)
Comment (innat) [reply keras_cv/layers/preprocessing/grayscale.py]: Done.

File: keras_cv/layers/preprocessing/grayscale.py (hunk 6 above)
Comment (innat) [reply keras_cv/layers/preprocessing/grayscale.py]: Done.

File: keras_cv/layers/preprocessing/grayscale.py (hunk 7 above)
Comment (innat) [reply keras_cv/layers/preprocessing/grayscale.py]: Done.

File: keras_cv/layers/preprocessing/grayscale.py (hunk 9 above)
Comment (innat) [reply keras_cv/layers/preprocessing/grayscale.py]: Done.

File: keras_cv/layers/preprocessing/grayscale.py (hunk 10 above)
Comment (innat) [reply keras_cv/layers/preprocessing/grayscale.py]: Done.

File: keras_cv/layers/preprocessing/grayscale.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+            images: Tensor representing images of shape
+                [batch_size, width, height, channels], with dtype tf.float32 / tf.uint8, or,
+                [width, height, channels], with dtype tf.float32 / tf.uint8
+        Returns:
+            images: augmented images, same shape as input.
+        """
+        if training is None:
+            training = backend.learning_phase()
+
+        return tf.__internal__.smart_cond.smart_cond(
```
[the remark this answers is not in the document]
Comment (innat) [reply keras_cv/layers/preprocessing/grayscale.py]: Done.

File: keras_cv/layers/preprocessing/grayscale.py (hunk 5 above)
Comment (innat) [reply keras_cv/layers/preprocessing/grayscale.py]: Done.

**Comment by LukeWood** [comment]
Thanks @innat - great contribution!!!
